staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/openapi_controller.go

```diff
@@ -0,0 +1,167 @@
+	AddUpdateApiService(handler http.Handler, apiService *apiregistration.APIService) error
+	UpdateApiServiceSpec(apiServiceName string, spec *spec.Swagger, etag string) error
+	RemoveApiServiceSpec(apiServiceName string) error
+	GetApiServiceInfo(apiServiceName string) (handler http.Handler, etag string, exists bool)
+}
+
+// OpenAPIAggregationController periodically check for changes in OpenAPI specs of APIServices and update/remove
+// them if necessary.
+type OpenAPIAggregationController struct {
+	openAPIAggregationManager OpenAPIAggregationManager
+	queue                     workqueue.RateLimitingInterface
+	downloader                *openAPIDownloader
+
+	// To allow injection for testing.
+	syncHandler func(key string) error
+}
+
+func NewOpenAPIAggregationController(downloader *openAPIDownloader, openAPIAggregationManager OpenAPIAggregationManager) *OpenAPIAggregationController {
+	c := &OpenAPIAggregationController{
+		openAPIAggregationManager: openAPIAggregationManager,
+		queue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.NewItemExponentialFailureRateLimiter(successfulUpdateDelay, failedUpdateMaxExpDelay), "APIServiceOpenAPIAggregationControllerQueue1"),
+		downloader:  downloader,
+	}
+
+	c.syncHandler = c.sync
+
+	return c
+}
+
+func (c *OpenAPIAggregationController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting OpenAPIAggregationController")
+	defer glog.Infof("Shutting down OpenAPIAggregationController")
+
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *OpenAPIAggregationController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *OpenAPIAggregationController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	defer c.queue.Done(key)
+	if quit {
+		return false
+	}
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("loading OpenAPI spec for %q failed with: %v", key, err))
+
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *OpenAPIAggregationController) sync(key string) error {
+	handler, etag, exists := c.openAPIAggregationManager.GetApiServiceInfo(key)
+	if !exists || handler == nil {
+		c.queue.Forget(key)
+		return nil
+	}
+	returnSpec, newEtag, httpStatus, err := c.downloader.Download(handler, etag)
+	rateLimit := false
+	switch {
+	case err != nil:
+		return err
+	case httpStatus == http.StatusNotModified:
+		rateLimit = false
+	case httpStatus == http.StatusNotFound || returnSpec == nil:
+		rateLimit = true
+	case httpStatus == http.StatusOK:
+		if err := c.openAPIAggregationManager.UpdateApiServiceSpec(key, returnSpec, newEtag); err != nil {
+			utilruntime.HandleError(fmt.Errorf("aggregating OpenAPI spec for %q failed with: %v", key, err))
+			rateLimit = true
+		}
+	}
+
+	if rateLimit {
+		c.queue.AddRateLimited(key)
```

Comment:
This is basically your error case. Just return an error and let the `c.queue.AddRateLimited` do its work in processNextWorkItem.